### PR TITLE
Improvement In AbstractOperation and Some Point Operations

### PR DESCRIPTION
**Pull request by AshishS-1123**
# Summary 

## Changes in Abstract Operation

- **set_parameter** - This method will be used for providing metadata for the operation. 
Usually, operations or brushes need some parameters for processing. Since each operation will have a bunch of different parameters, it will be hard for developers to keep track of the names of those parameters and their setters. This method gives us a convenient and universal way to set parameters.
- **set_operation_bounds** - This method will specify the region of the image over which to apply the operation.

## Changes in Image

**Image** has a number of new convenience functions for
- Cloning Images
- Creating Empty Images
- Getting and Setting Values of Specific Pixels

## Point Operation

Point operations are those operations that calculate the new value of a pixel based on its previous value. In this PR, I have developed point operations for **Brightness**, **Contrast** and **Inversion**.
This mainly serve as proof-of-concept for the upgrades to AbstractOperation, and also allow us to develop some tools in the App.

## Tests

I haven't made any major changes to how tests are organised or executed, but this PR should lay a few guidelines for how to write them.
Since we are adopting TDD, our approach will be to write tests first and lay down the functionality and then do the implementation.
Hence, the unittests must check that all the essential methods from **AbstractOperation** have been implemented (like the set_parameters and set_operation_bounds methods). This will ensure (both to the developer and to the reviewer) that the behaviour is being controlled by standard methods. It will also check the correctness of implementation (for example, when developing an operation, it is very easy to apply the operation on the whole image rather than the operation bounds.).

# Testing Changes

- Run ```meson build`` from root directory. This step need not be done for every compile.
- Then ```cd build`` to go to the build directory.
- Then run ```ninja test``` to run all the unittests. This will show you the names of tests that were successful and otherwise.

### CODE REVIEW

**Review by colinkiama — APPROVED** [review]

**Review by colinkiama — COMMENTED** [review]
When we get a good amount of tests, it would be nice to add test completion results for each test unit too

File: src/PointOps/BrightnessContrast.vala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                uchar red = (uchar) ((pixel[0] - 128) * contrast + brightness + 128);
+                uchar green = (uchar) ((pixel[1] - 128) * contrast + brightness + 128);
+                uchar blue = (uchar) ((pixel[2] - 128) * contrast + brightness + 128);
+                uchar alpha = pixel[3];
+
+                // Make sure that channel values lie within accepted range.
+                Utils.Maths.clamp_pixel (ref red);
+                Utils.Maths.clamp_pixel (ref green);
+                Utils.Maths.clamp_pixel (ref blue);
+                processed.set_pixel(i, j, red, green, blue, alpha);
```
Comment (ghost) [inline src/PointOps/BrightnessContrast.vala]:
```suggestion
                processed.set_pixel (i, j, red, green, blue, alpha);
```

File: src/PointOps/BrightnessContrast.vala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    public override void set_parameter (string parameter, string value) {
+        switch (parameter) {
+            case "brightness":
+                brightness = double.parse (value);
+                break;
+            case "contrast":
+                contrast = double.parse (value);
+                break;
+            default:
+                warning ("""Invalid parameter %s given to BrightnessContrast operation.""".printf(parameter));
```
Comment (ghost) [inline src/PointOps/BrightnessContrast.vala]:
```suggestion
                warning ("""Invalid parameter %s given to BrightnessContrast operation.""".printf (parameter));
```

File: src/PointOps/Inversion.vala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                var blue = 255 - pixel[2];
+                var alpha = pixel[3];
+
+                processed.set_pixel (i, j, red, green, blue, alpha);
+            }
+        }
+    }
+
+    public override void set_parameter (string parameter, string value) {
+        warning ("""Invalid parameter %s given to Inversion operation.""".printf(parameter));
```
Comment (ghost) [inline src/PointOps/Inversion.vala]:
```suggestion
        warning ("""Invalid parameter %s given to Inversion operation.""".printf (parameter));
```

File: tests/PointOps/test_Inversion.vala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    inv_op.original = new LibPhotostat.Image.from_size (10, 10);
+
+    var bounds = LibPhotostat.Utils.Rectangle (3, 3, 8, 8);
+    inv_op.set_operation_bounds (bounds);
+
+    assert (inv_op.operation_bounds == bounds);
+}
+
+// Tests that setting bounds and performing operation only affects the concerned region
+void test_bounds_effect ()  {
```
Comment (ghost) [inline tests/PointOps/test_Inversion.vala]:
```suggestion
void test_bounds_effect () {
```

**Review by ghost — CHANGES_REQUESTED** [review]
Some formatting fixes

File: src/Image.vala [hunk 5]
```diff
@@ -20,5 +20,52 @@
  */
 
 public class LibPhotostat.Image {
+    public enum FORMAT {
```
Comment (ghost) [inline src/Image.vala:23]: Shouldn't this be `static` ?

Comment (AshishS-1123) [reply src/Image.vala:23]: @Abdallah-Moh **enum** already has static like behaviour. You can access it by referencing the class name as such
```vala
LibPhotostat.Image.FORMAT.RGB
```
So there is no need for the **static** keyword. Adding the keyword gives syntax error anyway.

File: src/Utils/Maths.vala [hunk 6]
```diff
@@ -0,0 +1,11 @@
+public class LibPhotostat.Utils.Maths {
+    public static void clamp_pixel (ref uchar value) {
+        if (value > 255) {
+            value = 255;
+        }
+
+        if (value < 0) {
+            value = 0;
+        }
```
Comment (ghost) [inline src/Utils/Maths.vala]: No need for an if statement just use
```suggestion
        value = value > 255 ? 255 : (value < 0 ? 0 : value)
```

Comment (AshishS-1123) [reply src/Utils/Maths.vala]: @Abdallah-Moh I will commit this suggestion. But doesn't this make the code a little harder to read?

Comment (ghost) [reply src/Utils/Maths.vala]: A comment will do the trick

**Pull request merged**